## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [14]:
import pandas as pd
from pathlib import Path

df = pd.read_csv('Resources/client_dataset.csv')

df.head()


first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [15]:
# View the column names in the data
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [16]:
# Use the describe function to gather some basic statistics

# print(df.describe())
df.describe()



client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [17]:
# Use this space to do any additional research
# and familiarize yourself with the data.
# df.dtypes
# df.index
# df.max
# df.count() 
# df['client_id'].value_counts().head(10)
df['client_id'].unique()
# df['client_id'].nunique()
# df['client_id'].value_counts().head(10)


array([58515, 37609, 57113, 46554, 92089, 33615, 87213, 46709, 73362,
       58789, 81012, 74112, 18145, 48855, 67407, 26185, 17355, 28523,
       77184, 61603, 51869, 85413, 74159, 33804, 54932, 87125, 10643,
       10351, 10559, 31733, 95847, 97623, 68703, 28927, 48015, 90654,
       67516, 75118, 32185, 96736, 89499, 79827, 45569, 61647, 53409,
       17002, 18349, 25936, 37715, 41908, 25397, 59472, 77583, 34507,
       17697, 33956, 56645, 42481, 28011, 21654, 43361, 73729, 32597,
       34878, 24053, 48148, 27983, 55398, 40622, 41766, 40677, 92169,
       74058, 79361, 31268, 34313, 95108, 34706, 70043, 12286, 76087,
       99043, 28331, 47644, 13436, 31165, 79451, 22890, 14180, 32098,
       71958, 56590, 84527, 38220, 19206, 91091, 14312, 75546, 57762,
       10520, 86810, 31715, 65049, 87506, 22670, 49058, 82075, 37608,
       96954, 53348, 40490, 97794, 22008, 77340, 47507, 86045, 99231,
       49037, 19820, 61465, 48193, 32620, 14935, 16444, 37542, 46820,
       16500, 65084,

In [18]:
print("qty:")
print(f"The count is {df['qty'].count()}")
print(f"The minimum is {df['qty'].min()}")
print(f"The maximum is {df['qty'].max()}")
print(f"The mean is {df['qty'].mean()}")

qty:
The count is 54639
The minimum is 0
The maximum is 3958244
The mean is 570.2646095279928


In [19]:

# print(df['client_id'])
# print(len(df['client_id']))
# print(df["category"].unique())
# print(df["category"].nunique())
print(df["category"].value_counts())

category
consumables    23538
furniture      11915
software        8400
decor           7954
electronics     2832
Name: count, dtype: int64


In [20]:
# What three item categories had the most entries?
df.sort_values(by='category', ascending=False)
df.groupby('category').sum()
df.groupby('category').size().sort_values(ascending=False).head(3)


category
consumables    23538
furniture      11915
software        8400
dtype: int64

In [21]:
# For the category with the most entries,
# which subcategory had the most entries?

df.sort_values(by='category', ascending=False)
df.groupby('category').sum()
df.groupby('category').size().sort_values(ascending=False).head(3)
df.groupby('subcategory').size().sort_values(ascending=False).head(1)


subcategory
bathroom supplies    6424
dtype: int64

In [22]:
# Which five clients had the most entries in the data?
df.sort_values(by='client_id', ascending=False)
df.groupby('client_id').sum()
df.groupby('client_id').size().sort_values(ascending=False).head(5)


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
dtype: int64

In [23]:
# Store the client ids of those top 5 clients in a list.
Top5_clients = df.groupby('client_id').size().sort_values(ascending=False).head(5).index.tolist()
print(Top5_clients)



[33615, 66037, 46820, 38378, 24741]


In [35]:
# How many total units (the qty column) did the
# client with the most entries order order?
# total_units = df.groupby('client_id')['qty'].sum().sort_values(ascending=False)
# print(total_units.head(1))

df.loc[df['client_id'] == 33615, 'qty'].sum()    



64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [25]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df['subtotal'] = df['unit_price'] * df['qty']
# print(df['subtotal'])
print(df.head())



       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor  

In [49]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
# df['weight'] = df['qty'] * df['unit_weight']


# df['shipping_price'] = 
#     if df[weight] > 50:
#         return weight * 7
#     else:
#         return weight * 10
# print(df.head())
#create a dataframe

# df = df.apply(shipping_price)    
# print(df)
# print(shipping_price(5))
# print(df.head())

# df['shipping_price'] = df['qty'] * df['unit_weight']*7 when over 50 pounds
# df['shipping_price'] = df['qty'] * df['unit_weight']*10 when 50 pounds or under
# print(df['shipping_price'])
# import numpy as np
# df['shipping_price'] = np.where(df['unit_weight'] > 50, 
#                                 df['qty'] * df['unit_weight'] * 7, 
#                                 df['qty'] * df['unit_weight'] * 10)
# print(df)
def calculate_shipping_price(weight):
    if weight > 50:
        return weight * 7
    else:
        return weight * 10

df['shipping_price'] = df['weight'].apply(calculate_shipping_price)
print(df.head())


       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            w

In [51]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%

df['total_price'] = (df['subtotal'] + df['shipping_price'])*1.0925    
print(df.head())     

# df['total_price'] = (df['subtotal'] + df['shipping_price'])*1.0925
# df['total_price'] = (df['subtotal'] + ('shipping_price'))*1.0925
# print(df['total_price'],['total_price_no_tax'])



       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.

In [53]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['total_cost'] = ((df['unit_cost'] * df['qty'])+df['shipping_price'])
print(df['total_cost'])
print(df.head())    


0        85597.05
1          629.79
2          765.18
3         1077.35
4         8663.60
           ...   
54634     2222.55
54635    12095.92
54636    41054.25
54637    14280.00
54638    16599.10
Name: total_cost, Length: 54639, dtype: float64
       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com 

In [54]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total_price'] - df['total_cost'] 
print(df['profit']) 
print(df.head())    




0        46242.026250
1          284.468625
2          311.937600
3          466.221400
4         2700.803500
             ...     
54634     1342.321200
54635     2717.320275
54636     8130.646250
54637     1352.713600
54638     3112.331250
Name: profit, Length: 54639, dtype: float64
       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913  had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [59]:
# Check your work using the totals above
# Determine total price + shopping price, sum and display for order id 2742071
# df['total_Price_nt'] = ((df['unit_price'] * df['qty'])+df['shipping_price'])
# order_total = df.loc[df['order_id'] == 6128929, 'total_price'].sum() + df.loc[df['order_id'] == 6128929, 'shipping_price'].sum()
# Determine total price + shiping + tax for order id 2742071
order_total = df.loc[df['order_id'] == 2742071, 'total_price'].sum()
print(order_total)







152811.897075


In [60]:
# Determine total price + shiping + tax for order id 2173913
order_total = df.loc[df['order_id'] == 2173913, 'total_price'].sum()
print(order_total)

162388.7193


In [61]:
# Determine total price + shiping + tax for order id 6128929
order_total = df.loc[df['order_id'] == 6128929, 'total_price'].sum()
print(order_total)

923441.24045


In [62]:
print(df.head())    



       first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_cost unit_weight  qty  line_number  \
0          17        2023  ...    762.71        7.50  105   

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [78]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.

# df.groupby('client_id')['total_price'].sum().sort_values(ascending=False).head(5)
# print(df.groupby('client_id')['total_price'].sum().sort_values(ascending=False).head(5))
# df.loc[df['client_id'] in 33615, 'total_price'].sum()    
# Top5_clients

# values = Top5_clients
# Top5_clients_sales = df.loc[df['client_id'].isin(Top5_clients)].groupby('client_id')['total_price'].sum()   
# Top5_clients_sales.sort_values('Total Profit', ascending=False)
# print(Top5_clients_sales)

Top5_clients_sales = df.loc[df['client_id'].isin(Top5_clients)].groupby('client_id')['total_price'].sum()
Top5_clients_sales = Top5_clients_sales.sort_values(ascending=False)
for client_id, sales in Top5_clients_sales.items():
    print(f"Client ID: {client_id}, Total Sales: {sales:,.2f}" )

Client ID: 24741, Total Sales: 82,268,891.98
Client ID: 38378, Total Sales: 12,906,550.88
Client ID: 66037, Total Sales: 10,259,514.80
Client ID: 46820, Total Sales: 9,743,794.32
Client ID: 33615, Total Sales: 8,377,308.57


In [94]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.

# data_dict = {'client_id': Top5_clients}
# columns = ['qty', 'shipping_price', 'total_price', 'total_shipping_price', 't   otal_revenue', 'total_profit'] 
# for col in columns:
#     data_dict[col] = [total
# summary_df


# Convert Top5_clients list into a DataFrame
# Ensure 'qty' column exists in 'Top5_clients_df'
# if 'qty' not in Top5_clients_df.columns:
#     Top5_clients_df['qty'] = 0

# # Create pivot table
# summary_df = Top5_clients.pivot_table(
#     index='client_id',
#     values=['qty', 'shipping_price', 'total_price', 'total_profit'],
#     aggfunc='sum'
# )

# # Sort by total profit
# summary_df = summary_df.sort_values('total_profit', ascending=False)


# Ryan gave hint pivot tables could be used here, utilized GPT to assist with troubelshooting the code
# Define Top5_clients_df as a DataFrame
Top5_clients_df = df[df['client_id'].isin(Top5_clients)]

# Add 'qty' column if it doesn't exist
if 'qty' not in Top5_clients_df.columns:
    Top5_clients_df['qty'] = 0

# Create pivot table
summary_df = Top5_clients_df.pivot_table(
    index='client_id',
    values=['qty', 'shipping_price', 'total_price', 'profit'],
    aggfunc='sum'
)

print(summary_df)


                 profit     qty  shipping_price   total_price
client_id                                                    
24741      3.657999e+07  239862      5126448.37  8.226889e+07
33615      2.201995e+06   64313      1828984.89  8.377309e+06
38378      3.271830e+06   73667      3429455.40  1.290655e+07
46820      2.736603e+06   75768      1601448.84  9.743794e+06
66037      3.255032e+06   43018      1395151.85  1.025951e+07


In [103]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.
# pivot_df = pivot_df.rename(columns={'total_price': 'Total Sales', 'qty': 'Quantity'})

# Convert DataFrame values to numbers
summary_df = summary_df.apply(pd.to_numeric, errors='coerce')

# Define a function to format values
def format_values(x):
    return "${:,.2f}M".format(x / 1000000)

# Apply the function to the pivot table
summary_df = summary_df.applymap(format_values)

print(summary_df)
summary_df['Total Sales'] = summary_df['total_price'].apply(format_values)
summary_df['Quantity'] = summary_df['Quantity'].apply(format_values)
print(summary_df)



          profit    qty shipping_price total_price
client_id                                         
24741      $nanM  $nanM          $nanM       $nanM
33615      $nanM  $nanM          $nanM       $nanM
38378      $nanM  $nanM          $nanM       $nanM
46820      $nanM  $nanM          $nanM       $nanM
66037      $nanM  $nanM          $nanM       $nanM


C:\Users\KURTD\AppData\Local\Temp\ipykernel_26828\2368864874.py:14: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  summary_df = summary_df.applymap(format_values)


TypeError: unsupported operand type(s) for /: 'str' and 'int'

In [ ]:
# Sort the updated data by "Total Profit" form highest to lowest
